mesonbuild/coredata.py

```diff
@@ -902,6 +905,19 @@ def _parse_section(self, s):
                 raise EnvironmentException(f'Malformed value in machine file variable {entry!r}.')
             except KeyError as e:
                 raise EnvironmentException(f'Undefined constant {e.args[0]!r} in machine file variable {entry!r}.')
+
+            if s == 'binaries':
+                if isinstance(res, list):
+                    for element in res:
+                        if not isinstance(element, str):
+                            raise EnvironmentException(f'Invalid type {res!r} for entry {entry!r} in machine file')        
+                elif not isinstance(res, str):
+                    raise EnvironmentException(f'Invalid type {res!r} for entry {entry!r} in machine file')
```

Comment:
You can use `stringlistify()` that also checks all items are strings.

Reply:
`stringlistify()` would result in a more general exception message, while the code above points the developer to his error.
With regard to clean code, `stringlistify()` should definitely be preferred.